Verify `HostTensorDescriptor` when it is created

emezh opened the pull request:
## Proposed changes

Verify tensor descriptor is valid, specifically tensor dimensions vs strides.

For example, a 128x128 matrix can't have a stride of 64 in the outer dimension.

### Issues
[LWPCK-3425](https://ontrack-internal.amd.com/browse/LWPCK-3425) [gemm_multiply_add] accuracy issue observed with multiple tests	
[LWPCK-3417](https://ontrack-internal.amd.com/browse/LWPCK-3417) [gemm_fastgelu] accuracy issue observed with multiple tests	
[LWPCK-3405](https://ontrack-internal.amd.com/browse/LWPCK-3405) [gemm_add_relu] accuracy issue observed with multiple tests	
[LWPCK-3403](https://ontrack-internal.amd.com/browse/LWPCK-3403) [gemm_add_relu_add_layernorm] accuracy issue observed with multiple tests

In all the issues above, ckProfiler is invoked with invalid parameters and these are not checked, hence the incorrect results.

### The Fix
The fix is to verify tensor descriptor (`HostTensorDescriptor`) when it is created using a new `HostTensorDescriptor::ValidateStrides`.

PR #2638 did something similar, but requires calling validate function explicitly. The solution in this PR is more generic and, essentially, it reverts changes in PR #2638.

### Other (new) Features
1. A specific layout (one of in `ck::tensor_layout`) must be passed to constructor `HostTensorDescriptor`. The default value for layout is `ck::tensor_layout::BaseTensorLayout` and, in most cases, it will cause a validation error (with a few exceptions - keep reading).
2. For legacy GEMM and element-wise tests that often pass just tensor dimensions (1D or 2D) and no or partially initialized strides, the layout will be assumed to be RowMajor or ColumnMajor based on stride values. See `HostTensorDescriptor::HandleDefaultLayout` and `HostTensorDescriptor::CalculateStrides`
3. `HostTensorDescriptor::CalculateStrides` is enhanced to detect partially initialized strides for GEMM layouts. This is done mostly for legacy GEMM cases to avoid modifying many existing GEMM tests to pass RowMajor/ColumnMajor explicitly (70+ files). 
4. To streamline operations in `HostTensorDescriptor` ctors, they are refactored to:
- a single master constructor 
- the rest of specialized constructors are made delegating constructors

### Future Work
- Fix first 8 test configs in _test/gemm_multi_abd/test_gemm_multi_abd_xdl.cpp_ - `HostTensorDescriptor` validation throws an error.
- `HostTensorDescriptor::ValidateStrides` is currently implemented for GEMM Layouts only and needs to be updated for Convolution Layouts - [here](https://github.com/ROCm/composable_kernel/blob/89ec40d2bac9ca35a97002abf868f6dc33e3196d/include/ck/library/utility/host_tensor.hpp#L439)
- transpose functions in [convolution_host_tensor_descriptor_helper.hpp](https://github.com/ROCm/composable_kernel/pull/2829/files#diff-72f9434eb6254d59db46e4f4b2034b99eb8b3a1b750b6f92d0adcc5494e58a2e) will need to be updated accordingly.
 - `transpose_host_tensor_descriptor_given_new2old` may need to be updated too - [here](https://github.com/ROCm/composable_kernel/pull/2829/files#diff-72f9434eb6254d59db46e4f4b2034b99eb8b3a1b750b6f92d0adcc5494e58a2eR206)

### Manual tests
LWPCK-3425:
Manual tests for profile_gemm_multiply_add (matrix B in Row and Col modes)
```
auto-strides
	bin/ckProfiler gemm_multiply_add 0 0 1 1 0 1 128 128 128 0 0 0 0 0
	bin/ckProfiler gemm_multiply_add 0 1 1 1 0 1 128 128 128 0 0 0 0 0
	bin/ckProfiler gemm_multiply_add 0 0 1 1 0 1 128 128 128 -1 -1 -1 -1 -1
Note, -1 should be deprecated (use 0 instead)

explicit strides (same as auto)
	bin/ckProfiler gemm_multiply_add 0 0 1 1 0 1 128 128 128 128 128 128 128 128
	bin/ckProfiler gemm_multiply_add 0 1 1 1 0 1 128 128 128 128 128 128 128 128

explicit strides (not the same as auto)
	bin/ckProfiler gemm_multiply_add 0 0 1 1 0 1 128 128 128 130 132 134 136 138
	bin/ckProfiler gemm_multiply_add 0 1 1 1 0 1 128 128 128 130 132 134 136 138

mix of explicit and auto strides
	bin/ckProfiler gemm_multiply_add 0 0 1 1 0 1 128 128 128 128 128 128 128 0

invalid stride
	bin/ckProfiler gemm_multiply_add 0 0 1 1 0 1 128 128 128 0 0 0 0 64
	terminate called after throwing an instance of 'std::runtime_error'
	  what():  Invalid strides for RowMajor: mLens: 128 128 , mStrides: 64 1 
	Aborted (core dumped)
```

LWPCK-3417, LWPCK-3405, LWPCK-3403: 
```
root@rocm-ck-eugene-77b75fdd47-8vzgz:~/git_internal/composable_kernel/build# bin/ckProfiler gemm_fastgelu 1 0 1 2 0 1 128 128 128 64 64 64
terminate called after throwing an instance of 'std::runtime_error'
  what():  Invalid strides for RowMajor: dim 2, lengths {128, 128}, strides {64, 1}
Aborted (core dumped)
```
## Checklist

Please put an `x` into the boxes that apply. You can also fill these out after creating the PR. If you're not sure, please don't hesitate to ask.

- [ ] I have added tests relevant to the introduced functionality, and the unit tests are passing locally
- [ ] I have added the test to REGRESSION_TESTS list defined at the top of CMakeLists.txt in tests/CMakeLists.txt, **IF** the test takes more than 30 seconds to run.
- [x] I have added inline documentation which enables the maintainers with understanding the motivation
- [ ] I have removed the stale documentation which is no longer relevant after this pull request
- [ ] (If this change is user-facing) I have added release notes which provide the end users with a brief summary of the improvement from this pull request
- [x] I have run `clang-format` on all changed files
- [ ] Any dependent changes have been merged

emezh committed:
first draft

emezh committed:
add proper GEMM layout verification

emezh committed:
clang-format

rahjain-amd commented on include/ck/library/utility/host_tensor.hpp:
```diff
@@ -99,56 +101,166 @@ auto construct_f_unpack_args(F, T args)
+                         )
+        : mLens(std::move(lens)), mStrides(std::move(strides))
+    {
+        if(mStrides.empty())
```
I am not sure, If strides will be empty with ckProfiler, because its must to pass the input in the command like, -1 means it will calculate the stride itself. 

<img width="658" height="258" alt="image" src="https://github.com/user-attachments/assets/c2b6d8ce-74df-4ee7-934f-ec505201bb91" />

emezh replied:
Good point.
Is it correct to assume that ROCm doesn't support negative strides?
Hypothetically, it enables to access the tensor in reverse order, i.e. by passing a pointer to the end of the tensor and specifying negative strides. If ROCm doesn't support such a mode, then stride=-1 can be handled in `HostTensorDescriptor` constructor properly.

Some tests (if not all?) check explicitly for -1 in strides and set them to positive value, e.g. 

https://github.com/ROCm/composable_kernel/blob/321627aec5ba27ec6c96ac44e8c9d82d26a5944c/profiler/src/profile_gemm_multiply_add.cpp#L95

and then

https://github.com/ROCm/composable_kernel/blob/321627aec5ba27ec6c96ac44e8c9d82d26a5944c/profiler/src/profile_gemm_multiply_add.cpp#L119

Besides, `CalculateStrides` (which currently handles `mStrides.empty()`) assumes RowMajor order.

Bottom line, there are two options here:
A. [simple] let the caller (i.e. the test or example code) handle negative strides and in `HostTensorDescriptor` just assert that none of the strides is negative.  `CalculateStrides` will be changed to assert on layout being `RowMajor` (will need to see if there are convolution examples that pass empty strides. If so, will need to handle that...

B. [more complicated] scratch `CalculateStrides` and handle empty `mStrides` or negative stride values in `ValidateStrides` based on the layout. That will work under assumption that negative strides are generally not supported by ROCm.

Any preference based on past experience? I'm more inclined to go with option A.

emezh replied:
`mStrides` are defined as `std::vector<std::size_t>` so negative strides are not supported and, strictly speaking, can't be detected.
In most of the cases, ckProfiler tests cast `int Stride` to `size_t` in the helper functions, e.g. 
```
    auto f_host_tensor_descriptor =
        [](std::size_t row, std::size_t col, std::size_t stride, auto layout) {
                                             ^^^^^^^^^^^ implicit cast to unsigned
            ...
        };

    Tensor<ADataType> a_m_k(f_host_tensor_descriptor(M, K, StrideA, ALayout{}));
                                                           ^^^^^^^ defined as int
```
Passing -1 as-is will yield `stride=UINT_MAX` and eventually lead to segmentation fault. Proper fix will require a lot of refactoring in the existing code. For the purpose of this PR, I'll do a hack of converting strides to `int` (temporarily in a local var) in `ValidateStrides` and assert they are positive.

ThomasNing replied:
I think Option A is better. Also, one issue we have it now is sometimes we put stride=0 as default sometimes we put stride = -1 as default. Let's unify that.

emezh replied:
If we want a simple solution, then for strides=0 we can check if all of them are zero and then call `CalculateStrides` (which will assert RowMajor only). If only some of the strides are zero, then handling becomes layout-dependent and quite elaborate - essentially this is option B.

emezh replied:
@rahjain-amd this is addressed in [b5f5f52](https://github.com/ROCm/composable_kernel/pull/2829/commits/b5f5f52fba0114b688861b90097704af628ff0e6) - pls take a look. I've also update PR description wrt to what has been tested. Note, that passing 0 or -1 to ckProfiler tests does require explicit handling - see the changes to profile_gemm_multiply_add.hpp and .cpp

ThomasNing commented on include/ck/library/utility/host_tensor.hpp:
```diff
@@ -99,56 +101,166 @@ auto construct_f_unpack_args(F, T args)
+        assert(mLens.size() == mStrides.size());
+
+        if(mLens.empty())
+            // TBD: should we throw error for 0-dim tensor?
```
There should be no 0-dim tensor. We could throw a static assert here.

ThomasNing commented on include/ck/tensor_operation/gpu/device/tensor_layout.hpp:
```diff
@@ -27,6 +28,24 @@ struct MFMA : public BaseTensorLayout
+{
+};
+template <>
+struct is_gemm_layout<MFMA> : std::true_type
```
Layout MFMA is a special operation that we will preshuffle the gemm into the shape of Warp GEMM. Let's omit it right now and focus on the RowMajor and ColumnMajor.

emezh replied:
I think this declaration is still needed for the sake of completeness. What we can do is to avoid throwing in MFMA case in `ValidateStrides`:

https://github.com/ROCm/composable_kernel/blob/0f056fc375e5b253eba2db59313e66f6821315c8/include/ck/library/utility/host_tensor.hpp#L175

ThomasNing reviewed (commented):
In general, it looks good to me. Some comments on the content.

emezh committed:
Handle "auto" strides.

CalculateStrides only called when tensor's strides are empty or all of them are <=0 (auto strides).
CalculateStrides now supports GEMM::ColumnsMajor order. The assumption is still that it applies only to the inner two dims.
ValidateStrides throws if any of the tensor's strides is <=0.
profile_gemm_multiply_add updated to support "auto" strides for tensors.

Manual tests for profile_gemm_multiply_add (matrix B in Row and Col modes)
auto-strides
	bin/ckProfiler gemm_multiply_add 0 0 1 1 0 1 128 128 128 0 0 0 0 0
	bin/ckProfiler gemm_multiply_add 0 1 1 1 0 1 128 128 128 0 0 0 0 0
	bin/ckProfiler gemm_multiply_add 0 0 1 1 0 1 128 128 128 -1 -1 -1 -1 -1
Note, -1 should be deprecated (use 0 instead)

explicit strides (same as auto)
	bin/ckProfiler gemm_multiply_add 0 0 1 1 0 1 128 128 128 128 128 128 128 128
	bin/ckProfiler gemm_multiply_add 0 1 1 1 0 1 128 128 128 128 128 128 128 128

explicit strides (not the same as auto)
	bin/ckProfiler gemm_multiply_add 0 0 1 1 0 1 128 128 128 130 132 134 136 138
	bin/ckProfiler gemm_multiply_add 0 1 1 1 0 1 128 128 128 130 132 134 136 138

mix of explicit and auto strides
	bin/ckProfiler gemm_multiply_add 0 0 1 1 0 1 128 128 128 128 128 128 128 0

invalid stride
	bin/ckProfiler gemm_multiply_add 0 0 1 1 0 1 128 128 128 0 0 0 0 64
	terminate called after throwing an instance of 'std::runtime_error'
	  what():  Invalid strides for RowMajor: mLens: 128 128 , mStrides: 64 1
	Aborted (core dumped)

ThomasNing commented on include/ck/library/utility/host_tensor.hpp:
```diff
@@ -99,56 +101,204 @@ auto construct_f_unpack_args(F, T args)
-    void CalculateStrides();
+        if constexpr(std::is_same_v<ck::tensor_layout::BaseTensorLayout, Layout>)
+        {
+            // TBD: should we throw error instead of warning?
```
I think we could first throw error here.

ThomasNing commented on include/ck/library/utility/host_tensor.hpp:
```diff
@@ -99,56 +101,204 @@ auto construct_f_unpack_args(F, T args)
+#endif
+        else
+        {
+            // TBD: if desired, a new "bypass" layout can be added to skip the error, e.g.
```
Great idea.

ThomasNing commented on include/ck/library/utility/host_tensor.hpp:
```diff
@@ -99,56 +101,204 @@ auto construct_f_unpack_args(F, T args)
+            }
+        }
+#if 0
+        // TBD: is_convolution_layout is not implemented yet
```
Convolution is another circumstance for calculating the stride. You could ask for @bartekxk's advice on this part.

emezh committed:
- add more names to ck::tensor_layout for easier namespace hierarchy checking
- updated convolutional layouts to use explicit ones or BaseConvolutionalLayout where it is not clear which layout to use (TBD) - see include/ck/library/utility/convolution_host_tensor_descriptor_helper.hpp

emezh committed:
clang-format

emezh committed:
more test fixes

emezh committed:
added handling of partially initialized strides for GEMM. fixed more tests.

emezh committed:
clang-format and more fixes

illsilin committed:
Merge branch 'develop' into emezhibo/check_strides

emezh committed:
replace long dash by a simple hyphen - causes build failure in CK codegen.

emezh committed:
Merge branch 'emezhibo/check_strides' of github.com-public:ROCm/composable_kernel into emezhibo/check_strides

emezh committed:
increase sizeof input, otherwise output size becomes zero or negative with large filter size

emezh committed:
select stride based on layout

emezh committed:
specify layout explicitly to avoid errors in HostTensorDescriptor creation

emezh committed:
add validation for higher GEMM tensor dimensions.; Add docstring to `HostTensorDescriptor`

emezh committed:
Not clear why permute test in test/permute_scale/test_permute_scale.cpp uses a lot of invalid strides. Setting layout to BypassLayoutVerification to avoid a lot of errors

emezh committed:
another layout fix

emezh committed:
clang-format

emezh committed:
test hook

emezh committed:
clang format

emezh committed:
more test fixes

emezh committed:
more test fixes

emezh committed:
more test fixes

emezh committed:
more conv tests fixes

emezh committed:
fix test (incl removing invalid config)

emezh committed:
fix moe examples:
- (in .cpp) add layout argument to non-2D tensors
- (in .hpp) fix asserts/failures that show up in Debug mode, specifically addressing 2D tensor by a single index (and 3D tensor by 2d index)

emezh committed:
Merge branch 'develop' into emezhibo/check_strides

emezh committed:
fix test and a compile error.

emezh commented:
I'm still cleaning/updating the failing tests, but I'd like to start with a formal review. 

@ThomasNing pls tag anybody else who needs to take a look.
@rahjain-amd I think your change request has been addressed - pls take a look and if you feel that's true - pls accept or remove the request.

rahjain-amd reviewed (dismissed):
Looks good To me

emezh committed:
fix test_contraction

emezh committed:
fix moe_gemm2 example.

emezh committed:
fix profile and wmma examples

emezh committed:
clean-up early mods for ckprofile. verified with:
```
ckProfiler gemm_multiply_add 0 0 1 1 0 1 128 128 128 0 0 0 0 0
ckProfiler gemm_multiply_add 0 1 1 1 0 1 128 128 128 0 0 0 0 0
ckProfiler gemm_multiply_add 0 0 1 1 0 1 128 128 128 130 132 134 136 138
ckProfiler gemm_multiply_add 0 1 1 1 0 1 128 128 128 130 132 134 136 138
#
ckProfiler gemm_fastgelu 1 0 1 2 0 1 128 128 128 0 0 0
ckProfiler gemm_fastgelu 1 1 1 2 0 1 128 128 128 0 0 0
ckProfiler gemm_fastgelu 1 2 1 2 0 1 128 128 128 0 0 0
ckProfiler gemm_fastgelu 1 3 1 2 0 1 128 128 128 0 0 0
ckProfiler gemm_fastgelu 1 0 1 2 0 1 128 128 128 128 128 128
#
ckProfiler gemm_add_relu 0 0 1 1 0 1 128 128 128 0 0 0 0
# ckProfiler gemm_add_relu 0 1 1 1 0 1 128 128 128 0 0 0 0    # not implemented
# ckProfiler gemm_add_relu 0 2 1 1 0 1 128 128 128 0 0 0 0    # not implemented
# ckProfiler gemm_add_relu 0 3 1 1 0 1 128 128 128 0 0 0 0    # not implemented
ckProfiler gemm_add_relu 0 0 1 1 0 1 128 128 128 128 128 128 128
#
ckProfiler gemm_add_relu_add_layernorm 1 0 1 1 0 0 128 128 128 0 0 0 0 0
ckProfiler gemm_add_relu_add_layernorm 1 1 1 1 0 0 128 128 128 0 0 0 0 0
ckProfiler gemm_add_relu_add_layernorm 1 2 1 1 0 0 128 128 128 0 0 0 0 0
ckProfiler gemm_add_relu_add_layernorm 1 3 1 1 0 0 128 128 128 0 0 0 0 0
ckProfiler gemm_add_relu_add_layernorm 1 0 1 1 0 0 128 128 128 130 132 134 136 138
#
example_gemm_add_multiply_dl_fp16
example_gemm_add_multiply_xdl_fp16
#
ckProfiler gemm_blockscale_wp 7 1 1 1 1 0 1 128 128 128 0 0 0
ckProfiler gemm_blockscale_wp 7 1 1 1 1 0 1 128 128 128 128 128 128
```

emezh committed:
fix minor compile error

emezh committed:
turn off debug mode; replace some of the asserts by runtime error; clean-up; better comments, etc.

emezh committed:
Merge branch 'develop' into emezhibo/check_strides

emezh committed:
fix example/67_

emezh committed:
clang-format

emezh committed:
temporary skip first 8 test configs - they throw error

emezh committed:
temporary skip first 8 test configs in wmma too - they throw error
